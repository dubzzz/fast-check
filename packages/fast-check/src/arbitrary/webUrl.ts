import { constantFrom } from './constantFrom';
import { constant } from './constant';
import { option } from './option';
import { tuple } from './tuple';
import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { webQueryParameters } from './webQueryParameters';
import { webFragments } from './webFragments';
import type { WebAuthorityConstraints } from './webAuthority';
import { webAuthority } from './webAuthority';
import { partsToUrlMapper, partsToUrlUnmapper } from './_internals/mappers/PartsToUrl';
import type { SizeForArbitrary } from './_internals/helpers/MaxLengthFromMinLength';
import { relativeSizeToSize, resolveSize } from './_internals/helpers/MaxLengthFromMinLength';
import { webPath } from './webPath';

const safeObjectAssign = Object.assign;

/**
 * Constraints to be applied on {@link webUrl}
 * @remarks Since 1.14.0
 * @public
 */
export interface WebUrlConstraints {
  /**
   * Enforce specific schemes, eg.: http, https
   * @defaultValue ['http', 'https']
   * @remarks Since 1.14.0
   */
  validSchemes?: string[];
  /**
   * Settings for {@link webAuthority}
   * @defaultValue &#123;&#125;
   * @remarks Since 1.14.0
   */
  authoritySettings?: WebAuthorityConstraints;
  /**
   * Enable query parameters in the generated url
   * @defaultValue false
   * @remarks Since 1.14.0
   */
  withQueryParameters?: boolean;
  /**
   * Enable fragments in the generated url
   * @defaultValue false
   * @remarks Since 1.14.0
   */
  withFragments?: boolean;
  /**
   * Define how large the generated values should be (at max)
   * @remarks Since 2.22.0
   */
  size?: Exclude<SizeForArbitrary, 'max'>;
}

/**
 * For web url
 *
 * According to {@link https://www.ietf.org/rfc/rfc3986.txt | RFC 3986} and
 * {@link https://url.spec.whatwg.org/ | WHATWG URL Standard}
 *
 * @param constraints - Constraints to apply when building instances
 *
 * @remarks Since 1.14.0
 * @public
 */
export function webUrl(constraints?: WebUrlConstraints): Arbitrary<string> {
  const c = constraints || {};
  const resolvedSize = resolveSize(c.size);
  const resolvedAuthoritySettingsSize =
    c.authoritySettings !== undefined && c.authoritySettings.size !== undefined
      ? relativeSizeToSize(c.authoritySettings.size, resolvedSize)
      : resolvedSize;
  // TODO - Move back to object spreading as soon as we bump support from es2017 to es2018+
  const resolvedAuthoritySettings = safeObjectAssign(safeObjectAssign({}, c.authoritySettings), {
    size: resolvedAuthoritySettingsSize,
  });
  const validSchemes = c.validSchemes || ['http', 'https'];
  const schemeArb = constantFrom(...validSchemes);
  const authorityArb = webAuthority(resolvedAuthoritySettings);
  return tuple(
    schemeArb,
    authorityArb,
    webPath({ size: resolvedSize }),
    c.withQueryParameters === true ? option(webQueryParameters({ size: resolvedSize })) : constant(null),
    c.withFragments === true ? option(webFragments({ size: resolvedSize })) : constant(null),
  ).map(partsToUrlMapper, partsToUrlUnmapper);
}
