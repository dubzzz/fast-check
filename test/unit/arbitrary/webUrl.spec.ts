import fc from '../../../lib/fast-check';
import { webUrl, WebUrlConstraints } from '../../../src/arbitrary/webUrl';
import { URL } from 'url';

import {
  assertProduceCorrectValues,
  assertProduceSameValueGivenSameSeed,
  assertProduceValuesShrinkableWithoutContext,
  assertShrinkProducesSameValueWithoutInitialContext,
} from './__test-helpers__/NextArbitraryAssertions';
import { Value } from '../../../src/check/arbitrary/definition/Value';
import { buildNextShrinkTree, renderTree } from './__test-helpers__/ShrinkTree';

function beforeEachHook() {
  jest.resetModules();
  jest.restoreAllMocks();
  fc.configureGlobal({ beforeEach: beforeEachHook });
}
beforeEach(beforeEachHook);

describe('webUrl (integration)', () => {
  type Extra = WebUrlConstraints;
  const extraParameters: fc.Arbitrary<Extra> = fc.record(
    {
      validSchemes: fc.constant(['ftp']),
      authoritySettings: fc.record(
        {
          withIPv4: fc.boolean(),
          withIPv6: fc.boolean(),
          withIPv4Extended: fc.boolean(),
          withUserInfo: fc.boolean(),
          withPort: fc.boolean(),
        },
        { requiredKeys: [] }
      ),
      withQueryParameters: fc.boolean(),
      withFragments: fc.boolean(),
    },
    { requiredKeys: [] }
  );

  const isCorrect = (t: string) => {
    // Valid url given the specs defined by WHATWG URL Standard: https://url.spec.whatwg.org/
    // A TypeError will be thrown if the input is not a valid URL: https://nodejs.org/api/url.html#url_constructor_new_url_input_base
    expect(() => new URL(t)).not.toThrow();
  };

  const webUrlBuilder = (extra: Extra) => webUrl(extra);

  it('should produce the same values given the same seed', () => {
    assertProduceSameValueGivenSameSeed(webUrlBuilder, { extraParameters });
  });

  it('should only produce correct values', () => {
    assertProduceCorrectValues(webUrlBuilder, isCorrect);
  });

  it('should produce values seen as shrinkable without any context', () => {
    assertProduceValuesShrinkableWithoutContext(webUrlBuilder, { extraParameters });
  });

  it('should be able to shrink to the same values without initial context', () => {
    assertShrinkProducesSameValueWithoutInitialContext(webUrlBuilder, { extraParameters });
  });

  it.each`
    rawValue
    ${'http://my.domain.org/a/z'}
    ${'http://user:pass@my.domain.org/a/z'}
    ${'http://my.domain.org/a/z?query#fragments'}
  `('should be able to shrink $rawValue', ({ rawValue }) => {
    // Arrange
    const arb = webUrl({
      authoritySettings: { withUserInfo: true },
      withQueryParameters: true,
      withFragments: true,
    });
    const value = new Value(rawValue, undefined);

    // Act
    const renderedTree = renderTree(buildNextShrinkTree(arb, value, { numItems: 100 })).join('\n');

    // Assert
    expect(arb.canShrinkWithoutContext(rawValue)).toBe(true);
    expect(renderedTree).toMatchSnapshot();
  });
});
