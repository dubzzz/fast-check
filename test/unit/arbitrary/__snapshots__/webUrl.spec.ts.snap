// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webUrl (integration) should be able to shrink http://my.domain.org/a/z 1`] = `
"\\"http://my.domain.org/a/z\\"
├> \\"http://domain.org/a/z\\"
|  ├> \\"http://aomain.org/a/z\\"
|  |  ├> \\"http://a.org/a/z\\"
|  |  |  ├> \\"http://a.rg/a/z\\"
|  |  |  |  ├> \\"http://a.ag/a/z\\"
|  |  |  |  |  ├> \\"http://a.ag\\"
|  |  |  |  |  ├> \\"http://a.ag/z\\"
|  |  |  |  |  |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  └> \\"http://a.ag\\"
|  |  |  |  |  |  ├> \\"http://a.ag/a\\"
|  |  |  |  |  |  |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  └> \\"http://a.ag/\\"
|  |  |  |  |  |  |     └> \\"http://a.ag\\"
|  |  |  |  |  |  ├> \\"http://a.ag/n\\"
|  |  |  |  |  |  |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  ├> \\"http://a.ag/h\\"
|  |  |  |  |  |  |  |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  ├> \\"http://a.ag/e\\"
|  |  |  |  |  |  |  |  |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |  |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  ├> \\"http://a.ag/c\\"
|  |  |  |  |  |  |  |  |  |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |  ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |  |  |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |  └> \\"http://a.ag/b\\"
|  |  |  |  |  |  |  |  |  |     ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |     ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |  |     |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |     └> \\"http://a.ag/a\\"
|  |  |  |  |  |  |  |  |  |        ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |        └> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |  |           └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  └> \\"http://a.ag/d\\"
|  |  |  |  |  |  |  |  |     ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |     ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |     |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |     └> \\"http://a.ag/c\\"
|  |  |  |  |  |  |  |  |        ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |        ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |        |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |        ├> \\"http://a.ag/a\\"
|  |  |  |  |  |  |  |  |        |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |        |  └> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |        |     └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |        └> \\"http://a.ag/b\\"
|  |  |  |  |  |  |  |  |           ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |           ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |           |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |           └> \\"http://a.ag/a\\"
|  |  |  |  |  |  |  |  |              ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |              └> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |                 └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  └> \\"http://a.ag/g\\"
|  |  |  |  |  |  |  |     ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |     ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |     |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |     └> \\"http://a.ag/f\\"
|  |  |  |  |  |  |  |        ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |        ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |        |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |        └> \\"http://a.ag/e\\"
|  |  |  |  |  |  |  |           ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |           ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |           |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |           ├> \\"http://a.ag/a\\"
|  |  |  |  |  |  |  |           |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |           |  └> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |           |     └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |           ├> \\"http://a.ag/c\\"
|  |  |  |  |  |  |  |           |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |           |  ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |           |  |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |           |  └> \\"http://a.ag/b\\"
|  |  |  |  |  |  |  |           |     ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |           |     ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |           |     |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |           |     └> \\"http://a.ag/a\\"
|  |  |  |  |  |  |  |           |        ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |           |        └> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |           |           └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |           └> \\"http://a.ag/d\\"
|  |  |  |  |  |  |  |              ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |              ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |              |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |              └> \\"http://a.ag/c\\"
|  |  |  |  |  |  |  |                 ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |                 ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |                 |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |                 ├> \\"http://a.ag/a\\"
|  |  |  |  |  |  |  |                 |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |                 |  └> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |                 |     └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |                 └> \\"http://a.ag/b\\"
|  |  |  |  |  |  |  |                    └> …
|  |  |  |  |  |  |  └> …
|  |  |  |  |  |  └> …
|  |  |  |  |  └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;

exports[`webUrl (integration) should be able to shrink http://my.domain.org/a/z?query#fragments 1`] = `
"\\"http://my.domain.org/a/z?query#fragments\\"
├> \\"http://domain.org/a/z?query#fragments\\"
|  ├> \\"http://aomain.org/a/z?query#fragments\\"
|  |  ├> \\"http://a.org/a/z?query#fragments\\"
|  |  |  ├> \\"http://a.rg/a/z?query#fragments\\"
|  |  |  |  ├> \\"http://a.ag/a/z?query#fragments\\"
|  |  |  |  |  ├> \\"http://a.ag?query#fragments\\"
|  |  |  |  |  |  ├> \\"http://a.ag#fragments\\"
|  |  |  |  |  |  |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  ├> \\"http://a.ag#ments\\"
|  |  |  |  |  |  |  |  ├> \\"http://a.ag#nts\\"
|  |  |  |  |  |  |  |  |  ├> \\"http://a.ag#ts\\"
|  |  |  |  |  |  |  |  |  |  ├> \\"http://a.ag#s\\"
|  |  |  |  |  |  |  |  |  |  |  ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  ├> \\"http://a.ag#a\\"
|  |  |  |  |  |  |  |  |  |  |  |  └> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  ├> \\"http://a.ag#j\\"
|  |  |  |  |  |  |  |  |  |  |  |  ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  ├> \\"http://a.ag#f\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"http://a.ag#d\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\"http://a.ag#c\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |     ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"http://a.ag#b\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |        ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |        └> \\"http://a.ag#a\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |           └> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  └> \\"http://a.ag#e\\"
|  |  |  |  |  |  |  |  |  |  |  |  |     ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"http://a.ag#d\\"
|  |  |  |  |  |  |  |  |  |  |  |  |        ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |        ├> \\"http://a.ag#a\\"
|  |  |  |  |  |  |  |  |  |  |  |  |        |  └> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |        └> \\"http://a.ag#c\\"
|  |  |  |  |  |  |  |  |  |  |  |  |           ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |           └> \\"http://a.ag#b\\"
|  |  |  |  |  |  |  |  |  |  |  |  |              ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |              └> \\"http://a.ag#a\\"
|  |  |  |  |  |  |  |  |  |  |  |  |                 └> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  ├> \\"http://a.ag#h\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  └> \\"http://a.ag#g\\"
|  |  |  |  |  |  |  |  |  |  |  |  |     ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"http://a.ag#f\\"
|  |  |  |  |  |  |  |  |  |  |  |  |        ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |        ├> \\"http://a.ag#a\\"
|  |  |  |  |  |  |  |  |  |  |  |  |        |  └> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |        ├> \\"http://a.ag#d\\"
|  |  |  |  |  |  |  |  |  |  |  |  |        |  ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |        |  └> \\"http://a.ag#c\\"
|  |  |  |  |  |  |  |  |  |  |  |  |        |     ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |        |     └> \\"http://a.ag#b\\"
|  |  |  |  |  |  |  |  |  |  |  |  |        |        ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |        |        └> \\"http://a.ag#a\\"
|  |  |  |  |  |  |  |  |  |  |  |  |        |           └> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |        └> \\"http://a.ag#e\\"
|  |  |  |  |  |  |  |  |  |  |  |  |           ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |           └> \\"http://a.ag#d\\"
|  |  |  |  |  |  |  |  |  |  |  |  |              ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |              ├> \\"http://a.ag#a\\"
|  |  |  |  |  |  |  |  |  |  |  |  |              |  └> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |              └> \\"http://a.ag#c\\"
|  |  |  |  |  |  |  |  |  |  |  |  |                 ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |                 └> \\"http://a.ag#b\\"
|  |  |  |  |  |  |  |  |  |  |  |  |                    ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |                    └> \\"http://a.ag#a\\"
|  |  |  |  |  |  |  |  |  |  |  |  |                       └> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |  └> \\"http://a.ag#i\\"
|  |  |  |  |  |  |  |  |  |  |  |     ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |     └> \\"http://a.ag#h\\"
|  |  |  |  |  |  |  |  |  |  |  |        ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |        ├> \\"http://a.ag#a\\"
|  |  |  |  |  |  |  |  |  |  |  |        |  └> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |        ├> \\"http://a.ag#e\\"
|  |  |  |  |  |  |  |  |  |  |  |        |  ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |        |  ├> \\"http://a.ag#c\\"
|  |  |  |  |  |  |  |  |  |  |  |        |  |  ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |        |  |  └> \\"http://a.ag#b\\"
|  |  |  |  |  |  |  |  |  |  |  |        |  |     ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |        |  |     └> \\"http://a.ag#a\\"
|  |  |  |  |  |  |  |  |  |  |  |        |  |        └> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |        |  └> \\"http://a.ag#d\\"
|  |  |  |  |  |  |  |  |  |  |  |        |     ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |        |     └> \\"http://a.ag#c\\"
|  |  |  |  |  |  |  |  |  |  |  |        |        ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |        |        ├> \\"http://a.ag#a\\"
|  |  |  |  |  |  |  |  |  |  |  |        |        |  └> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |        |        └> \\"http://a.ag#b\\"
|  |  |  |  |  |  |  |  |  |  |  |        |           ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |        |           └> \\"http://a.ag#a\\"
|  |  |  |  |  |  |  |  |  |  |  |        |              └> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |        └> \\"http://a.ag#g\\"
|  |  |  |  |  |  |  |  |  |  |  |           ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |           └> \\"http://a.ag#f\\"
|  |  |  |  |  |  |  |  |  |  |  |              ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |              └> \\"http://a.ag#e\\"
|  |  |  |  |  |  |  |  |  |  |  |                 ├> \\"http://a.ag#\\"
|  |  |  |  |  |  |  |  |  |  |  |                 ├> \\"http://a.ag#a\\"
|  |  |  |  |  |  |  |  |  |  |  |                 |  └> …
|  |  |  |  |  |  |  |  |  |  |  |                 └> …
|  |  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  └> …
|  |  |  |  |  |  └> …
|  |  |  |  |  └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;

exports[`webUrl (integration) should be able to shrink http://user:pass@my.domain.org/a/z 1`] = `
"\\"http://user:pass@my.domain.org/a/z\\"
├> \\"http://my.domain.org/a/z\\"
|  ├> \\"http://domain.org/a/z\\"
|  |  ├> \\"http://aomain.org/a/z\\"
|  |  |  ├> \\"http://a.org/a/z\\"
|  |  |  |  ├> \\"http://a.rg/a/z\\"
|  |  |  |  |  ├> \\"http://a.ag/a/z\\"
|  |  |  |  |  |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  ├> \\"http://a.ag/z\\"
|  |  |  |  |  |  |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  ├> \\"http://a.ag/a\\"
|  |  |  |  |  |  |  |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  └> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |     └> \\"http://a.ag\\"
|  |  |  |  |  |  |  ├> \\"http://a.ag/n\\"
|  |  |  |  |  |  |  |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  ├> \\"http://a.ag/h\\"
|  |  |  |  |  |  |  |  |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |  |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  ├> \\"http://a.ag/e\\"
|  |  |  |  |  |  |  |  |  |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |  ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |  |  |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |  ├> \\"http://a.ag/c\\"
|  |  |  |  |  |  |  |  |  |  |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |  |  ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |  |  |  |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |  |  └> \\"http://a.ag/b\\"
|  |  |  |  |  |  |  |  |  |  |     ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |  |     ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |  |  |     |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |  |     └> \\"http://a.ag/a\\"
|  |  |  |  |  |  |  |  |  |  |        ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |  |        └> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |  |  |           └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |  └> \\"http://a.ag/d\\"
|  |  |  |  |  |  |  |  |  |     ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |     ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |  |     |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |     └> \\"http://a.ag/c\\"
|  |  |  |  |  |  |  |  |  |        ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |        ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |  |        |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |        ├> \\"http://a.ag/a\\"
|  |  |  |  |  |  |  |  |  |        |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |        |  └> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |  |        |     └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |        └> \\"http://a.ag/b\\"
|  |  |  |  |  |  |  |  |  |           ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |           ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |  |           |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |           └> \\"http://a.ag/a\\"
|  |  |  |  |  |  |  |  |  |              ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  |              └> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |  |                 └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |  └> \\"http://a.ag/g\\"
|  |  |  |  |  |  |  |  |     ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |     ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |     |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |     └> \\"http://a.ag/f\\"
|  |  |  |  |  |  |  |  |        ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |        ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |        |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |        └> \\"http://a.ag/e\\"
|  |  |  |  |  |  |  |  |           ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |           ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |           |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |           ├> \\"http://a.ag/a\\"
|  |  |  |  |  |  |  |  |           |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |           |  └> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |           |     └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |           ├> \\"http://a.ag/c\\"
|  |  |  |  |  |  |  |  |           |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |           |  ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |           |  |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |           |  └> \\"http://a.ag/b\\"
|  |  |  |  |  |  |  |  |           |     ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |           |     ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |           |     |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |           |     └> \\"http://a.ag/a\\"
|  |  |  |  |  |  |  |  |           |        ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |           |        └> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |           |           └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |           └> \\"http://a.ag/d\\"
|  |  |  |  |  |  |  |  |              ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |              ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |              |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |              └> \\"http://a.ag/c\\"
|  |  |  |  |  |  |  |  |                 ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |                 ├> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |                 |  └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |                 ├> \\"http://a.ag/a\\"
|  |  |  |  |  |  |  |  |                 |  ├> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |                 |  └> \\"http://a.ag/\\"
|  |  |  |  |  |  |  |  |                 |     └> \\"http://a.ag\\"
|  |  |  |  |  |  |  |  |                 └> …
|  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  └> …
|  |  |  |  |  |  └> …
|  |  |  |  |  └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;
